src/server/editorServices.ts

```diff
@@ -1105,187 +1312,159 @@ namespace ts.server {
-                if (!scriptInfo || !project.isRoot(scriptInfo)) {
-                    rootFilesChanged = true;
-                    if (!scriptInfo) {
-                        const scriptKind = propertyReader.getScriptKind(f);
-                        const hasMixedContent = propertyReader.hasMixedContent(f, this.hostConfiguration.extraFileExtensions);
-                        scriptInfo = this.getOrCreateScriptInfoForNormalizedPath(normalizedPath, /*openedByClient*/ false, /*fileContent*/ undefined, scriptKind, hasMixedContent);
-                    }
-                }
-                newRootScriptInfos.push(scriptInfo);
-                newRootScriptInfoMap.set(scriptInfo.fileName, scriptInfo);
-            }
-
-            if (rootFilesChanged || newRootScriptInfos.length !== oldRootScriptInfos.length) {
-                let toAdd: ScriptInfo[];
-                let toRemove: ScriptInfo[];
-                for (const oldFile of oldRootScriptInfos) {
-                    if (!newRootScriptInfoMap.contains(oldFile.fileName)) {
-                        (toRemove || (toRemove = [])).push(oldFile);
+                let scriptInfo: ScriptInfo | NormalizedPath;
+                let path: Path;
+                if (!project.lsHost.fileExists(newRootFile)) {
+                    path = normalizedPathToPath(normalizedPath, this.currentDirectory, this.toCanonicalFileName);
+                    const existingValue = projectRootFilesMap.get(path);
+                    if (isScriptInfo(existingValue)) {
+                        project.removeFile(existingValue);
                     }
+                    projectRootFilesMap.set(path, normalizedPath);
+                    scriptInfo = normalizedPath;
                 }
-                for (const newFile of newRootScriptInfos) {
-                    if (!project.isRoot(newFile)) {
-                        (toAdd || (toAdd = [])).push(newFile);
-                    }
-                }
-                if (toRemove) {
-                    for (const f of toRemove) {
-                        project.removeFile(f);
-                    }
-                }
-                if (toAdd) {
-                    for (const f of toAdd) {
-                        if (f.isScriptOpen() && isRootFileInInferredProject(f)) {
+                else {
+                    const scriptKind = propertyReader.getScriptKind(f);
+                    const hasMixedContent = propertyReader.hasMixedContent(f, this.hostConfiguration.extraFileExtensions);
+                    scriptInfo = this.getOrCreateScriptInfoForNormalizedPath(normalizedPath, /*openedByClient*/ clientFileName === newRootFile, /*fileContent*/ undefined, scriptKind, hasMixedContent);
+                    path = scriptInfo.path;
+                    // If this script info is not already a root add it
+                    if (!project.isRoot(scriptInfo)) {
+                        project.addRoot(scriptInfo);

```

Comment:
It looks like `addRoot` already contains a check for `isRoot` -- maybe that's no longer necessary and could be an assertion?